test/Test/GameDAO.hs

```diff
@@ -0,0 +1,125 @@
+
+test_GameDAO :: TestTree
+test_GameDAO = testGroup "GameDAO Tests"
+  [ testGroup "Proposal creator:"
+      [ nettestScenario "can propose a valid proposal" validProposal
+      ]
+  , testGroup "Admin"
+      [ nettestScenario "can flush proposals that got accepted" flushAcceptedProposals
+      ]
+  ]
+
+validProposal :: (Monad m) => NettestImpl m -> m ()
+validProposal = uncapsNettest $ do
+  ((owner1, _), _, dao, _) <- originateBaseDaoWithConfig config
+
+  -- Fail due to proposing new content require 50 token.
+  callFrom (AddressResolved owner1) dao (Call @"Propose") (DAO.ProposeParams
+    { ppFrozenToken = 20
+    , ppProposalMetadata = sampleMetadataContent
+    }) & expectCustomError_ #fAIL_PROPOSAL_CHECK
+
+  callFrom (AddressResolved owner1) dao (Call @"Propose") (DAO.ProposeParams
+    { ppFrozenToken = 15
+    , ppProposalMetadata = sampleMetadataBalance
+    })
+
+  checkTokenBalance (DAO.frozenTokenId) dao owner1 15
+  checkTokenBalance (DAO.unfrozenTokenId) dao owner1 85
+
+flushAcceptedProposals :: (Monad m) => NettestImpl m -> m ()
+flushAcceptedProposals = uncapsNettest $ do
+  ((owner1, _), (owner2, _), dao, admin) <- originateBaseDaoWithConfig config
+
+  -- | Accepted Proposals
+  key1 <- createSampleProposal sampleMetadataBalance owner1 dao
+  checkTokenBalance (DAO.frozenTokenId) dao owner1 10
+
+  callFrom (AddressResolved admin) dao (Call @"Set_voting_period") 60
+  callFrom (AddressResolved admin) dao (Call @"Set_quorum_threshold") 1
+
+  let upvote = DAO.VoteParam
+        { vVoteType = True
+        , vVoteAmount = 2
+        , vProposalKey = key1
+        }
+      downvote = DAO.VoteParam
+        { vVoteType = False
+        , vVoteAmount = 1
+        , vProposalKey = key1
+        }
+  callFrom (AddressResolved owner2) dao (Call @"Vote") [upvote, downvote]
+  checkTokenBalance (DAO.frozenTokenId) dao owner2 3
+  checkTokenBalance (DAO.unfrozenTokenId) dao owner2 97
+
+  advanceTime (sec 61)
+  callFrom (AddressResolved admin) dao (Call @"Flush") ()
+
+  checkTokenBalance (DAO.frozenTokenId) dao owner1 0
+  checkTokenBalance (DAO.unfrozenTokenId) dao owner1 100 -- proposer
+
+  checkTokenBalance (DAO.frozenTokenId) dao owner2 0
+  checkTokenBalance (DAO.unfrozenTokenId) dao owner2 100 -- voter
+
+-------------------------------------------------------------------------------
+-- Helper
+-------------------------------------------------------------------------------
+sampleMetadataBalance :: GameDaoProposalMetadata
+sampleMetadataBalance = GameDaoProposalMetadata
+  { pmProposalType = BalanceType BalanceChange
+      { bcItemChanges = [ ItemChange
+          { icItemName = [mt|Ganto|]
+          , icChangelogs =
+              [ [mt|Ganto now has 10 charges.|]
+              , [mt|Ganto can now target enemey to reduce health regen.|]
+              ]
+          }]
+      , bcHeroChanges = []
+      }
+  , pmProposalDescription = [mt|Balance Item|]
+  }
```

Comment:
:pogchamp:

I recently asked @fersel whether can we have some gamedev-related project, and looks like here we have such partially.